---
title: "scRNAseq_QC_template"
author: "Emory Gencore"
date: '`r format(file.info("scRNAseq_QC_template.Rmd")$ctime, "%Y-%m-%d")`'
output: 
  rmdformats::robobook:
    fig_width: 8
    gallery: true
    thumbnails: false
---

<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r, include=FALSE}
library(Seurat)
library(tidyverse)
library(reshape2)
library(SingleR)
library(ape)
library(MAST)
library(glmGamPoi)
#library(package = affyLib, )
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, cache=TRUE, cache.lazy = FALSE)#, include=FALSE)
config <- yaml::yaml.load_file("scRNA_GEX_config.yml")
source('scRNAseq_functions.R')
# recluster <- FALSE
# resave_combined_obj <- FALSE
```


# QC

## Sequencing metrics

```{r}
samples <- read.table('samplesheet.txt', header=TRUE)
qc_metrics <- capture_QC_metrics(samples)
DT::datatable(qc_metrics, options = list(searching=FALSE, pageLength=20, lengthMenu=NULL))
```

```{r}
a <- c("Estimated Number of Cells", "Mean Reads per Cell", "Median Genes per Cell", "Number of Reads", "Reads Mapped Confidently to Genome", "Fraction Reads in Cells", "Total Genes Detected", "Median UMI Counts per Cell")
qc_metrics_plot <- gsub(",", "", qc_metrics)
qc_metrics_plot <- gsub("%", "", qc_metrics_plot)
b <- rownames(qc_metrics)
qc_metrics_plot <- as.data.frame(qc_metrics_plot)
rownames(qc_metrics_plot) <- b 
qc_metrics_plot[] <- sapply(qc_metrics_plot, as.numeric)#%>% mutate_all(function(x) as.numeric(as.character(x)))
ggplot(melt(t(qc_metrics_plot[rownames(qc_metrics_plot) %in% a,])), aes(x=Var1, y=value, fill=Var1)) + geom_col()+ facet_wrap(vars(Var2 ), ncol = 4, scales='free_y') + theme_bw() + labs(x=NULL, y=NULL, fill="Sample") #+ scale_y_continuous(limits=c(0,NA))
```


```{r peng_data, include=FALSE}
#, eval=!preran}
## generate the captures, pass in parameters based on your dataset
captures <- generate_captures(samples, min.cells=0, min.features=0, mt_pattern="^mt-", rb_pattern= "^Rp[sl]")
```

## Data quality

Dashed lines indicate 95% credible intervals for including 98% of data

```{r}
#using bootstrapping to estimate bet cutoffs
cutoffs <- lapply(captures, qc_vln_plots, bootstrap=TRUE)
```

```{r, include=FALSE}
##edit cutoffs if desired 
## cutoffs[[1]][1,1] <- 0
#then apply
my_subset <- function(capture, cutoffs) {
  subset(capture, subset = nFeature_RNA > cutoffs$nfeature[1] & nFeature_RNA < cutoffs$nfeature[2] &
           percent.mito < cutoffs$mito[2] & nCount_RNA > cutoffs$ncount[1] & nCount_RNA < cutoffs$ncount[2] )
}
captures <- mapply(my_subset, captures, cutoffs)
# captures$`LysM-cre` <- subset(captures$`LysM-cre`, subset = nFeature_RNA > 400 & nFeature_RNA < 4500 & percent.mito < 10)
# captures$`Ptpn11-E76K-pos-LysM-cre` <- subset(captures$`Ptpn11-E76K-pos-LysM-cre`, subset = nFeature_RNA > 400 & nFeature_RNA < 7000 & percent.mito < 10)
```

Plotting highly expressed genes. We expect to see constitutively expressed genes (e.g. mitochondrial, ribosomal, etc).

```{r most_abundant_genes}
lapply(captures, plot_top_genes)
```

```{r}
saveRDS(captures, "captures.Rds")
```




