---
author: "Emory Gencore"
date: '`r format(file.info("p22090_Tiffany_QC.Rmd")$ctime, "%Y-%m-%d")`'
output: 
  rmdformats::robobook:
    fig_width: 8
    gallery: true
    thumbnails: false
---

<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r, include=FALSE}
library(Seurat)
library(tidyverse)
library(reshape2)
library(SingleR)
library(ape)
library(MAST)
library(glmGamPoi)
library(kableExtra)
#library(package = affyLib, )
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, cache=TRUE, cache.lazy = FALSE)#, include=FALSE)
config <- yaml::yaml.load_file("scRNA_GEX_config.yml")
source('scRNAseq_functions.R')
# recluster <- FALSE
# resave_combined_obj <- FALSE
preran=TRUE
```


# Sequencing metrics {.tabset}

```{r}
samples <- read.table('samplesheet.txt', header=TRUE)
qc_metrics <- capture_QC_metrics_hashed(samples)
```

## Summary

Antibody capture summary

```{r}
short <- qc_metrics %>% filter(Category=='Cells', Library.Type=='Antibody Capture') %>% select(ID, Metric.Name, Metric.Value)
cnames <- short$Metric.Value
short <- short %>% reshape(idvar="ID",timevar = "Metric.Name", direction='wide') 
short <- rename_with(short, ~ gsub("Metric.Value.", "", .x, fixed=TRUE))
knitr::kable(short) %>%
  kable_styling(full_width = F)
```

```{r, fig.width=12, fig.height=5}
qc_metrics_plots <- qc_metrics
colnames(qc_metrics_plots) <- gsub(".", " ", colnames(qc_metrics_plots), fixed=TRUE)
qc_metrics_plots$`Metric Value` <- gsub("%", "", qc_metrics_plots$`Metric Value`)
qc_metrics_plots$`Metric Value` <- gsub(",", "", qc_metrics_plots$`Metric Value`)
qc_metrics_plots$`Metric Value` <- as.numeric(qc_metrics_plots$`Metric Value`)

a <- c("Estimated number of cells", "Mean reads per cell", "Fraction reads in cells", "Median UMI counts per cell", "Antibody reads usable per cell", "Fraction antibody reads usable")
tmp <- qc_metrics_plots %>% filter(`Metric Name` %in% a)#%>% mutate_all(function(x) as.numeric(as.character(x)))
ggplot(tmp %>% select(ID, `Library Type`, `Metric Name`, `Metric Value`) %>% filter(`Library Type`=='Antibody Capture'), aes(x=ID, y=`Metric Value`, fill=ID)) + geom_col() + facet_wrap(vars(`Metric Name`), scales='free_y', ncol=3) + ggtitle('Antibody Capture')
```

Gene expression summary

```{r}
short <- qc_metrics %>% filter(Category=='Cells', Library.Type=='Gene Expression') %>% select(ID, Metric.Name, Metric.Value)
cnames <- short$Metric.Value
short <- short %>% reshape(idvar="ID",timevar = "Metric.Name", direction='wide') 
short <- rename_with(short, ~ gsub("Metric.Value.", "", .x, fixed=TRUE))
knitr::kable(short) %>%
  kable_styling(full_width = F)#, options = list(searching=FALSE, pageLength=20, lengthMenu=NULL))
```


```{r, fig.width=12, fig.height=5}
a <- c("Estimated number of cells", "Mean reads per cell", "Median genes per cell", "Number of reads", "Confidently mapped to genome", "Fraction reads in cells", "Median UMI counts per cell")
tmp <- qc_metrics_plots %>% filter(`Metric Name` %in% a)
ggplot(tmp %>% select(ID, `Library Type`, `Metric Name`, `Metric Value`) %>% filter(`Library Type`=='Gene Expression'), aes(x=ID, y=`Metric Value`, fill=ID)) + geom_col() + facet_wrap(vars(`Metric Name`), scales='free_y', ncol=3) + ggtitle('Gene Expression')
```

## Detailed

```{r}
short <- qc_metrics %>% filter(Category=='Library', Library.Type=='Antibody Capture')# %>% select(ID, Metric.Name, Metric.Value)
# cnames <- short$Metric.Value
# short <- short %>% reshape(idvar="ID",timevar = "Metric.Name", direction='wide') 
# short <- rename_with(short, ~ gsub("Metric.Value.", "", .x, fixed=TRUE))
knitr::kable(short)#, options = list(searching=FALSE, pageLength=20, lengthMenu=NULL))
```

```{r}
short <- qc_metrics %>% filter(Category=='Library', Library.Type=='Gene Expression')# %>% select(ID, Metric.Name, Metric.Value)
knitr::kable(short)#, options = list(searching=FALSE, pageLength=20, lengthMenu=NULL))
```


# Hash demultiplexing

Hash metadata

```{r}
## Need HTO and Label columns for premade demultiplexing functions
hash_table <- data.frame(
  'Label'=c('RAo18',
'RCn19',
'ROi18',
'RSi19',
'RTa19',
'Rta19',
'ROi18',
'RAo18',
'RCn19',
'Rsi19'),
  'ID'=c('98677',
'98678',
'98679',
'98680',
'98681',
'98681',
'98679',
'98677',
'98678',
'98680'),
  'Timepoint'=c(rep('Apr05', 5), rep('Apr20', 5)),
loaded=c(4223,
6940,
318,
1058,
14104,
22344,
10012,
12627,
3323,
3377)
  )
hash_table$HTO <- plyr::mapvalues(hash_table$Hashtag, from=c(98677,98678,98679,98680,98681), to=c('Hash1Human','Hash2Human','Hash3Human','Hash4Human','Hash5Human'))
knitr::kable(hash_table %>% select(-loaded))
```

```{r, include=FALSE}
if (preran){
  captures <- readRDS('unfiltered_captures.Rds')
} else {
  captures <- mapply(create_hashed_objects, samples$FileID, samples$Label, hash_table=hash_table, mito.pattern = "^MT", ribo.pattern="^RP[SL]")
  names(captures) <- samples$Label
}
```

Hashtag oligo abundance distributions, grouping cells by primary HTO identity

```{r, fig.width=9, fig.height=9}
lapply(captures, function(x) {RidgePlot(x, assay='HTO', ncol=3, features=rownames(x[['HTO']]))})
```


Demultiplexing results

```{r}
tmp <- lapply(captures, function(x) {table(x$hash.ID)[order(table(x$hash.ID))]})
tmp <- do.call(cbind, tmp)
knitr::kable(tmp[order(rownames(tmp)),])
```

uMAP of HTO data 


```{r}
lapply(captures, function(x) {
  tmp <- subset(x, idents = "Negative", invert = TRUE)
  DefaultAssay(tmp) <- "HTO"
  tmp <- ScaleData(tmp, features = rownames(tmp), verbose = FALSE)
  tmp <- RunPCA(tmp, features = rownames(tmp), approx = FALSE)
  tmp <- RunUMAP(tmp, dims = 1:(length(levels(x@meta.data$hash.ID))-2), perplexity = 100)#(length(levels(captures$Apr05_HH_5A@meta.data$hash.ID))-2))#, perplexity = 100)
  DimPlot(tmp)
})
```


# Feature Summary

Identification of singlets, doublets, and negative GEMs is based on HTO tagging.

```{r, results='hide', fig.keep='all'}
qc_plots <- function(cap){
  plot1 <- FeatureScatter(cap, feature1 = "nCount_RNA", feature2 = "percent.mito", group.by = "HTO_classification.global")  +
    theme(legend.position="none") + ggtitle(cap@project.name)
plot2 <- FeatureScatter(cap, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "HTO_classification.global") +   ggtitle(cap@project.name)
plot1 + plot2 
}

lapply(captures, qc_plots)

```

## Most common features

Plotting highly expressed genes. We expect to see constitutively expressed genes (e.g. mitochondrial, ribosomal, etc).

```{r most_abundant_genes, fig.width=10, fig.height=4}
plot_top_genes <- function(data, n=20){
  C <- data@assays$RNA@counts
  C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
  most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
  par(mar=c(5,12,4,1)+.1)
  boxplot(as.matrix(t(C[most_expressed, ])), las=1, cex=0.1, xlab = "% total count per cell", col = (scales::hue_pal())(20)[20:1], horizontal = TRUE, main=data@project.name) 
}
tmp <- lapply(captures, plot_top_genes)
```

```{r, eval=!preran}
saveRDS(captures, "unfiltered_captures.Rds")
preran=TRUE
```

