---
title: 'p22091_Peng_analysis'
author: "Emory Gencore"
date: '`r Sys.Date()`'
output: 
  rmdformats::robobook:
    fig_width: 8
    gallery: true
    thumbnails: false
    toc_depth: 1
---
    
<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r, include=FALSE}
library(Seurat)
library(tidyverse)
library(reshape2)
library(SingleR)
library(ape)
library(MAST)
library(glmGamPoi)
library(kableExtra)
library(SoupX)
library(gridExtra)
library(DoubletFinder)
#library(SC3)
library(rngtools)
library(ggrepel)
library(DT)
library(clustree)
#library(package = affyLib, )
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, cache=FALSE, cache.lazy = FALSE, fig.width = 5, fig.height = 5)#, include=FALSE)
config <- yaml::yaml.load_file("scRNA_GEX_config.yml")
source('scRNAseq_functions.R')
# recluster <- FALSE
# resave_combined_obj <- FALSE
preran=TRUE
```


```{r}
combined.obj <- readRDS('saved_rds/combined_obj.Rds')
```

# Reference based cell-type identification

```{r reference_data, include=FALSE}
# eval=!preran 
seurat_map <- function(obj, ref, normalization.method = "SCT",
                       reference.reduction="pca",
                       refdata = list(cell_ontolog_class = "cell_ontology_class"),
                       ndims=20){
  anchors <- FindTransferAnchors(
    reference = ref,
    query = obj,
    normalization.method = normalization.method,
    reference.reduction = reference.reduction,
    dims = 1:ndims
  )
  obj <- MapQuery(
    anchorset = anchors,
    query = obj,
    reference = ref,
    refdata = refdata
  )
  obj
}
## Load reference object
load('droplet_Spleen_seurat_tiss.Robj')
tiss <- UpdateSeuratObject(tiss)
tiss <- SCTransform(tiss, verbose=FALSE)
combined.obj <- seurat_map(combined.obj, ref=tiss)

```

```{r}
mapping_results <- table(combined.obj@meta.data$predicted.cell_ontolog_class, combined.obj@meta.data$orig.ident)
knitr::kable(mapping_results)
```

```{r, results='hide', fig.keep='all'}

plot_mapping_results <- function(data){
  ggplot(data@meta.data, aes(x=predicted.cell_ontolog_class, y=predicted.cell_ontolog_class.score)) + geom_boxplot() + theme_minimal() + labs(x="Predicted celltype", y="Prediction score", caption="Higher prediction score is more confident") + geom_jitter(alpha=0.1)
  
}
plot_mapping_results(combined.obj)
```

```{r}
DefaultAssay(combined.obj) <- 'RNA'
```



<!-- # Cross-capture conserved cluster markers -->
# Removing unwanted celltypes

Celltype removal is based on clustering and marker gene identification, not reference mapping

```{r, include=FALSE, eval=FALSE}
if (preran) {
  #clusters.markers <- readRDS('saved_rds/clusters_markers.Rds')
  conserved_markers_pos <- readRDS('saved_rds/conserved_markers.Rds')
} else {
  #clusters.markers <- FindAllMarkers(combined.obj, min.pct = 0.25, logfc.threshold = log2(1.5), assay = 'RNA')
  # conserved_markers <- lapply(levels(combined.obj$seurat_clusters), function(x) {
  #   a <- FindConservedMarkers(combined.obj, assay='RNA', grouping.var = 'orig.ident', ident.1 = x,
  #                             logfc.threshold = log2(1.5), min.pct = 0.25)
  #   a$cluster <- x
  #   return(a %>% rownames_to_column('Gene'))
  # })
  conserved_markers_pos <- lapply(levels(combined.obj$seurat_clusters), function(x) {
    a <- FindConservedMarkers(combined.obj, assay='RNA', grouping.var = 'orig.ident', ident.1 = x,
                              logfc.threshold = log2(1.5), min.pct = 0.25, only.pos = TRUE)
    a$cluster <- x
    return(a %>% rownames_to_column('Gene'))
  })
  #conserved_markers <- bind_rows(conserved_markers)
  conserved_markers_pos <- bind_rows(conserved_markers_pos)
  #saveRDS(clusters.markers, "saved_rds/clusters_markers.Rds")
  saveRDS(conserved_markers_pos, "saved_rds/conserved_markers_pos.Rds")
}



# for (i in 0:(length(conserved_markers)-1)){
#   conserved_markers[[i+1]]$cluster <- i
#   conserved_markers[[i+1]] <- conserved_markers[[i+1]] %>% rownames_to_column('Gene')
# }

```

```{r, include=FALSE, eval=FALSE}
tmp <- as.data.frame.matrix(table(combined.obj$seurat_clusters, combined.obj$orig.ident))
tmp <- 100*tmp/rowSums(tmp)
#tmp <- tmp %>% mutate(across(.cols= everything(), ~ 100*.x / sum(.x))) #, .names = "{.col}_percent"
knitr::kable(tmp, caption = 'Percentage of cluster from each capture', digits = 1)
```

```{r , fig.height=16, fig.width=16, include=FALSE, eval=FALSE}
conserved_markers_pos$avg_log2FC <- (conserved_markers_pos %>% select(contains('log2FC')) %>% transmute(avg_log2FC=rowMeans(.)))$avg_log2FC
top10 <- conserved_markers_pos %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = abs(avg_log2FC))
DoHeatmap(combined.obj, features = top10$Gene)
```

```{r, include = FALSE, eval = FALSE}
datatable(conserved_markers %>% mutate_if(is.numeric, ~round(., 2)) %>% filter(max_pval<0.05) %>% dplyr::select(cluster, Gene, avg_log2FC, max_pval, `LysM-cre_pct.1`, `LysM-cre_pct.2`, `Ptpn11-E76K-pos-LysM-cre_pct.1`, `Ptpn11-E76K-pos-LysM-cre_pct.2`), filter = 'top', rownames = FALSE, colnames=c("Cluster", "Gene", "Mean log2 fold-change across conditions in cluster", "Max Adj. p val", "% detected in\nLysm-cre in cluster", "% detected in Lysm-cre noutside cluster", 
                            "% detected in Ptpn11-E76K-pos in cluster", "% detected in Ptpn11-E76K-pos outside cluster"              ))
```

<!-- Genes plotted here were pulled from top conserved markers as well as genes used in Peng's previous inquiry -->

```{r, fig.width=6}
markers <- read.csv('markers.txt', header=TRUE)
for (celltype in c('B cells', 'Plasma cells')){
  genes <- markers %>% filter(Celltype==celltype)
  a <- DotPlot(combined.obj, features=genes$Gene, assay='RNA', col.max = 50) + ggtitle(celltype)#, col.min = 0, col.max = 5) #+ ggtitle(celltype) #+ theme(axis.text.x = element_text(angle = 45, hjust=1))
  print(a)
}
#DotPlot(combined.obj, features=markers$Gene)+ theme(axis.text.x = element_text(angle = 45, hjust=1))
```

<!-- # Cluster labels -->

<!-- Final clusters with some manually provided labels. These do not represent any definitive statement about members of the cluster, but represent a non-expert manual approximation based on the metadata and markers. The previous sections shows genes defining clusters and metadata which may allow a mouse spleen expert to better characterize these groupings.  -->

```{r , fig.width=10, fig.height=10}
cluster_tags <- c(
  "0",#. Cd8a+ T cells",    #0
  "1",#. Ly6c1+ Cd4+ T cells",
  "2",#. Cd4+ T cells",
  "3",#. Cd8a+ T cells",
  "4",#. Neutrophils",    
  "5",#. Monocytes/Macrophages",               #5
  "6",#. Neutrophils",
  "7",#. Basophil or Erythroid",
  "8",#. DC (Myeloid?)",
  "9",#. Plasmatoid DC?",          
  "10",#. Neutrophil/Eosinophil?",                   #10
  "11",#. NK cells",
  "12",#. Monocytes",
  "13",#. DC",
  "14. Plasma cells",
  "15. B cells"        #15
  )
combined.obj$labels <- "NA"
combined.obj$labels <- plyr::mapvalues(combined.obj$seurat_clusters, 
                                       from=0:15, to=cluster_tags)
DimPlot(combined.obj, group.by = "labels", label = TRUE, label.box=TRUE, repel=TRUE, cells.highlight = colnames(subset(combined.obj, subset = seurat_clusters %in% c('14', '15')))) + theme(legend.position="none") + ggtitle("Unwanted clusters")
```

```{r}
combined.obj.full <- combined.obj
combined.obj.removed <- subset(combined.obj, subset = seurat_clusters %in% c(14, 15), invert=FALSE)
combined.obj <- subset(combined.obj, subset = seurat_clusters %in% c(14, 15), invert=TRUE)

```

```{r, include=FALSE}
DefaultAssay(combined.obj) <- 'integrated'
combined.obj <- RunPCA(combined.obj)
ElbowPlot(combined.obj, ndims = 30)
```

```{r, include=FALSE}
npcs = 12
combined.obj <- FindNeighbors(combined.obj, dims=1:npcs, verbose=FALSE)
combined.obj <- RunUMAP(combined.obj, reduction='pca', dims=1:npcs)
```

```{r, fig.width=12, fig.height=6, include = FALSE}
testclustree <- FindClusters(combined.obj, resolution = c(0.1,0.2,0.3,0.4,0.5,0.6,0.7), verbose=FALSE)
clustree(testclustree) +
    guides(edge_colour = FALSE) + theme(legend.position = "top")
```

```{r}
res = 0.3

combined.obj <- FindClusters(combined.obj, resolution = res, verbose=FALSE)
DefaultAssay(combined.obj) <- 'RNA'
combined.obj <- NormalizeData(combined.obj)
combined.obj <- FindVariableFeatures(combined.obj)
combined.obj <- ScaleData(combined.obj)
```

# Metadata 

## Cluster membership {.tabset}
How much of each capture is in each cluster

```{r}
tmp <- as.data.frame.matrix(table(combined.obj$seurat_clusters, combined.obj$orig.ident))
knitr::kable(tmp, caption = 'Number of cells per cluster')
tmp <- tmp %>% mutate(across(.cols= everything(), ~ 100*.x / sum(.x))) #, .names = "{.col}_percent"
knitr::kable(tmp, caption = '% of total capture per cluster', digits = 1)
tmp <- as.data.frame.matrix(table(combined.obj$seurat_clusters, combined.obj$orig.ident))
tmp <- 100*tmp/rowSums(tmp)
#tmp <- tmp %>% mutate(across(.cols= everything(), ~ 100*.x / sum(.x))) #, .names = "{.col}_percent"
knitr::kable(tmp, caption = '% of cluster belonging to each capture', digits = 1)
```


## Mapping features

```{r, fig.width=8}
DimPlot(combined.obj, split.by = "orig.ident") + ggtitle("Capture")
```

```{r, fig.width = 8, fig.height = 4}
p1 <- DimPlot(combined.obj, group.by = "predicted.cell_ontolog_class") + ggtitle("Predicted cell ontology class")
p2 <- FeaturePlot(combined.obj, reduction='umap', features="predicted.cell_ontolog_class.score") + ggtitle("Class prediction score")
p1 + p2
```

Cell cycle is predicted using human cell cycle markers mapped to mouse genes, so they may not be the most accurate for mouse cells

```{r}
DimPlot(combined.obj, group.by = "Phase") + ggtitle("Cell cycle phase")
```

```{r, fig.width = 8, fig.height = 8}
FeaturePlot(combined.obj, reduction='umap', features=c('nCount_RNA', "nFeature_RNA", "percent.mito", "percent.ribo"))
```

# Cluster labeling

```{r, include=FALSE, eval=TRUE}
if (preran) {
  #clusters.markers <- readRDS('saved_rds/clusters_markers.Rds')
  conserved_markers_pos <- readRDS('saved_rds/conserved_markers_pos.Rds')
} else {
  #clusters.markers <- FindAllMarkers(combined.obj, min.pct = 0.25, logfc.threshold = log2(1.5), assay = 'RNA')
  # conserved_markers <- lapply(levels(combined.obj$seurat_clusters), function(x) {
  #   a <- FindConservedMarkers(combined.obj, assay='RNA', grouping.var = 'orig.ident', ident.1 = x,
  #                             logfc.threshold = log2(1.5), min.pct = 0.25)
  #   a$cluster <- x
  #   return(a %>% rownames_to_column('Gene'))
  # })
  conserved_markers_pos <- lapply(levels(combined.obj$seurat_clusters), function(x) {
    a <- FindConservedMarkers(combined.obj, assay='RNA', grouping.var = 'orig.ident', ident.1 = x,
                              logfc.threshold = log2(1.5), min.pct = 0.25, only.pos = TRUE)
    a$cluster <- x
    return(a %>% rownames_to_column('Gene'))
  })
  #conserved_markers <- bind_rows(conserved_markers)
  conserved_markers_pos <- bind_rows(conserved_markers_pos)
  #saveRDS(clusters.markers, "saved_rds/clusters_markers.Rds")
  saveRDS(conserved_markers_pos, "saved_rds/conserved_markers_pos.Rds")
}



# for (i in 0:(length(conserved_markers)-1)){
#   conserved_markers[[i+1]]$cluster <- i
#   conserved_markers[[i+1]] <- conserved_markers[[i+1]] %>% rownames_to_column('Gene')
# }

```



Heatmap of the top 5 gene identifiers for each cluster.

```{r, fig.height=16, fig.width=16}
conserved_markers_pos$avg_log2FC <- (conserved_markers_pos %>% select(contains('log2FC')) %>% transmute(avg_log2FC=rowMeans(.)))$avg_log2FC
top10 <- conserved_markers_pos %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = abs(avg_log2FC))
DoHeatmap(combined.obj, features = top10$Gene)
```

```{r}
datatable(conserved_markers_pos %>% mutate_if(is.numeric, ~round(., 2)) %>% filter(max_pval<0.05) %>% dplyr::select(cluster, Gene, avg_log2FC, max_pval, `LysM-cre_pct.1`, `LysM-cre_pct.2`, `Ptpn11-E76K-pos-LysM-cre_pct.1`, `Ptpn11-E76K-pos-LysM-cre_pct.2`), filter = 'top', rownames = FALSE, colnames=c("Cluster", "Gene", "Mean log2 fold-change across conditions in cluster", "Max Adj. p val", "% detected in\nLysm-cre in cluster", "% detected in Lysm-cre noutside cluster", 
                            "% detected in Ptpn11-E76K-pos in cluster", "% detected in Ptpn11-E76K-pos outside cluster"              ))
```

Expression in dotplots is scaled

```{r, fig.width=6}
markers <- read.csv('markers.txt', header=TRUE)

markers[] <- lapply(markers, str_trim)
markers <- markers %>% filter(!Celltype %in% c('B cells', 'Plasma cells'))
for (celltype in unique(markers$Celltype)){
  genes <- markers %>% filter(Celltype==celltype)
  a <- DotPlot(combined.obj, features=genes$Gene, assay='RNA', col.max = 50) + theme(axis.text.x = element_text(angle = 45, hjust=1))+ ggtitle(celltype)#, col.min = 0, col.max = 5) #+ ggtitle(celltype) #+ theme(axis.text.x = element_text(angle = 45, hjust=1))
  print(a)
}
#DotPlot(combined.obj, features=markers$Gene)+ theme(axis.text.x = element_text(angle = 45, hjust=1))
```

# Final cluster labels

```{r label_clusters, fig.width=10, fig.height=10}
cluster_tags <- c(
  "0. NT/CM CD8+ T cells",    #0
  "1. NT/CM CD4+ T cells",
  "2. EM CD4+ T cells",
  "3. Neutrophils",
  "4. EM CD8+ T cells",#Neutrophils",    
  "5. Monocytes/Macrophages",               #5
  "6. Plasmacytoid DC?",
  "7. Neutrophils",
  "8. DC (Myeloid?)",
  "9. Basophil or Erythroid",          
  "10. Neutrophil/Eosinophil?",                   #10
  "11. Monocytes",
  "12. NK cells",
  "13. DC"
  )
combined.obj$labels <- "NA"
combined.obj$labels <- plyr::mapvalues(combined.obj$seurat_clusters, 
                                       from=0:(length(unique(combined.obj$seurat_clusters))-1), to=cluster_tags)
DimPlot(combined.obj, group.by = "labels", label = TRUE, label.box=TRUE, repel=TRUE) + theme(legend.position="none") + ggtitle("Manual labels")
```

# Differential gene expression {.tabset}

DGE within clusters of interest. Each comparison has a tab with the following components:

* Number of cells from each condition
* Conserved markers for the cluster are shown as safety check to confirm selection and annotation (these are the same as the previous section, just specific to the cluster of interest)
* Differentially expressed genes tables


DGE tables have the following values:

* Adj. p val : Adjusted p-value, based on Bonferroni correction using all features in the dataset.
* Mean log2 fold change : log fold-change of the average expression between the two groups. **Positive values indicate that the feature is more highly expressed in the  LysM-cre group**
* pct.1 : The percentage of cells where the feature is detected in the first group
* pct.2 : The percentage of cells where the feature is detected in the second group
    
    
```{r}
per_cluster_DGE <- function(cluster_num){
  #cat('Number of cells from each condition')
  counts <- table((combined.obj@meta.data %>% filter(seurat_clusters == cluster_num))$orig.ident)
  
  #cat('Conserved cluster markers')
  consmarkers <- datatable(conserved_markers_pos %>% filter(cluster == cluster_num) %>% mutate_if(is.numeric, ~round(., 2)) %>% filter(max_pval<0.05) %>% dplyr::select(cluster, Gene, avg_log2FC, max_pval, `LysM-cre_pct.1`, `LysM-cre_pct.2`, `Ptpn11-E76K-pos-LysM-cre_pct.1`, `Ptpn11-E76K-pos-LysM-cre_pct.2`), filter = 'top', rownames = FALSE, colnames=c("Cluster", "Gene", "Mean log2 fold-change across conditions in cluster", "Max Adj. p val", "% detected in\nLysm-cre in cluster", "% detected in Lysm-cre noutside cluster", 
                              "% detected in Ptpn11-E76K-pos in cluster", "% detected in Ptpn11-E76K-pos outside cluster"              ), caption = 'Marker genes for cluster')
  
  markers <- FindMarkers(subset(combined.obj, subset = seurat_clusters == cluster_num), group.by = 'orig.ident', ident.1 = "LysM-cre", ident.2 = "Ptpn11-E76K-pos-LysM-cre", assay = 'RNA', min.pct = 0.25, verbose = FALSE, logfc.threshold = log2(1.5), test.use = 'MAST') %>% rownames_to_column('Gene')
  
  #cat('DGE')
  
  dge <- datatable(markers %>% mutate_if(is.numeric, ~round(., 4)) %>% filter(p_val_adj<0.05) %>% dplyr::select(Gene, avg_log2FC, p_val_adj, pct.1, pct.2), filter = 'top', rownames = FALSE, colnames=c("Gene", "Mean log2\nfold change", "Adj. p val", "% detected\nin LysM-cre", "% detected\nin Ptpn11-E76K-pos-LysM-cre"), caption = 'DGE within cluster')
  return(list(counts=counts, conserved_markers=consmarkers, dge=dge))
}
```


## 0. NT/CM CD8+ T cells

```{r}
results <- per_cluster_DGE(0)
results$counts
results$conserved_markers
results$dge
```


## 4. EM CD8+ T cells

```{r}
results <- per_cluster_DGE(4)
results$counts
results$conserved_markers
results$dge
```

## 3. Neutrophils

```{r}
results <- per_cluster_DGE(3)
results$counts
results$conserved_markers
results$dge
```

## 7. Neutrophils

```{r}
results <- per_cluster_DGE(7)
results$counts
results$conserved_markers
results$dge
```

## 10. Neutrophils/eosinophils (?)

```{r}
results <- per_cluster_DGE(10)
results$counts
results$conserved_markers
results$dge
```

## 6. Plasmacytoid DC(?)

```{r}
results <- per_cluster_DGE(6)
results$counts
results$conserved_markers
results$dge
```