---
title: "scRNAseq processing template"
author: "ENPRC Genomics Core"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
# library(Azimuth)      ## Can be tricky to install, singleR is easier
library(BiocParallel)
library(gencoreSC)
library(ggpubr)
library(here)
library(scDblFinder)
library(Seurat)
library(SingleR)
library(tidyverse)

knitr::opts_chunk$set(echo=FALSE, warning=TRUE, message=TRUE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE,
                      fig.align = 'center')#, include=FALSE)
set.seed(45)
```

```{}
This script is intended to perform single-cell dataset processing prior
to downstream analysis. This includes removing low quality cells, removing doublets,
examining ambient RNA contamination, checking cell-cycle effects, demultiplexing
hashed data, normalizing ADT data, and mapping reference annotations. Ideally,
the product of this script is a single Seurat object (either merged or integrated)
that is ready to use for 'analysis' such as differential gene expression,
cluster labeling, etc.

The workflow in this script relies heavily on the gencoreSC package (see
https://github.com/yerkes-gencore/gencoreSC for more information). 
```

# Load configuration

```{r}
## Disables some functionality for test
samplesheet <- read.table(here('config/samplesheet.txt'), header=TRUE)
if ('capture_31' %in% samplesheet$FileID){
  message('Your samplesheet file has template defaults. Edit "config/samplesheet.txt" to start your project')
  message('Waiting 10 seconds to make sure you see this! If this is a test run, ignore this, script will resume shortly')
  Sys.sleep(10)
}
config <- yaml::yaml.load_file(here("config/scRNA_GEX_config.yml"))
if (config$alignmentDir == 'tools/templates/example_data/sc_ADT+GEX+HTO'){
  message('Your config file has template defaults. Edit "config/scRNA_GEX_config.yml" to start your project')
  message('Waiting 10 seconds to make sure you see this! If this is a test run, ignore this, script will resume shortly')
  Sys.sleep(10)
}
```

# GEX processing

## soupX

```{}
Running soupX requires the unfiltered matrices as well as the filtered
outputs from Cellranger
```


```{r}
## Finding files for SoupX, could also manually specify
## sometimes the pattern is different based on cellranger version
soupx_files <- lapply(samplesheet$FileID, function(x){
  findSoupXFiles(file.path(config$rootDir, config$alignmentDir, x), filtered_mat_pattern = 'sample_feature_bc_matrix$')
})
names(soupx_files) <- samplesheet$Label
# Running soupx takes a while
sc <- lapply(soupx_files, function(x){
  runSoupX(x[['unfiltered_mat_path']], x[['filtered_mat_path']], x[['clusters_path']])
})

saveRDS(sc, here('saved_rds/report-02/soupx_results.Rds'))
## always a good idea to reload the saved object to check that it saved ok
sc <- readRDS(here('saved_rds/report-02/soupx_results.Rds'))
```

```{r}
extractSoupXContamEst <- function(sc){
  rho <- sc$fit$rhoEst
  rho_low <- sc$fit$rhoFWHM[1]
  rho_high <- sc$fit$rhoFWHM[2]
  return(list(rho_low = rho_low, rho = rho, rho_high = rho_high))
}
soupx_results <- data.table::rbindlist(lapply(sc, extractSoupXContamEst), idcol = 'Sample')
soupx_results
```

```{r, eval = FALSE}
## The replay plot function bugs with html_notebook,
## will only show in viewer window if you execute it from console
## https://github.com/rstudio/rstudio/issues/5648
## Pulling individual plots to examine possible problem samples
sc$R1$plot
sc$R4$plot
```

```{r, eval = FALSE}
## See highly affected genes
plotSoupXGeneAdjustments(sc$R1, hide_ensembl = TRUE, ens_pattern = '^ENS')
plotSoupXGeneAdjustments(sc$R4, hide_ensembl = TRUE, ens_pattern = '^ENS')
```

## Load data

```{}
depending on the results of soupX, you may decide to use the corrected counts
instead of the raw counts. In that case you will have to adjust this
workflow or overwrite the produced Seurat objects' data slots
```

```{r, include=FALSE}
objs <- mapply(readCounts10x,
               samplesheet$Label,
               file.path(config$rootDir,
                         config$alignmentDir, 
                         samplesheet$FileID,
                         'outs/per_sample_outs', 
                         samplesheet$FileID, 
                         'count/sample_feature_bc_matrix.h5'),
               strip.suffix = FALSE,
               format = 'h5',
               USE.NAMES = TRUE)

## Is this even necessary? 
# saveRDS(objs, here('saved_rds/raw_captures.Rds'))

```

```{r}
## Renaming cells with the sample name to avoid conflicts with references, merging etc.
for (sample in names(objs)){
  objs[[sample]]@project.name = paste0(sample)
}

objs <- lapply(objs, function(sample){
  RenameCells(sample, add.cell.id = sample@project.name)
})
```

```{r}
## This keeps track of dropped cells etc. through processing steps
qc_receipts <- list()
umaps.rna <- list()

qc_receipts$preprocessing <- lapply(objs, dim)
## can and should modify the dims manually
umaps.rna$dims <- c(ceiling(sqrt(length(objs))),ceiling(sqrt(length(objs))))
```

## Removing low count genes

Rather than using the min.cells argument of readCounts10x, I prefer to separate
this step out. This has the benefit of 1) allowing you to slot in the soupX
adjusted counts easily and 2) more explicitly seeing the impact of this step.
You may not need to do this if you are interested in low abundance genes,
but it can improve memory use and run speed for some things. 

```{r}
## only genes appearing in at least `min_cells` cells in N-1 samples will be kept
objs <- filterLowGenes(objs, min_cells = 50, assay = 'RNA')
## If we decide we don't want joint filtering, but on a per-capture basis, you
## can still use this function on individual objects
## objs <- lapply(objs, filterLowGenes)

qc_receipts$remove_low_genes <- lapply(objs, dim)
```

## Hash demultiplexing

Not needed for all studies

```{r}
HTO_table <- read.csv(here('config/HTO_table.csv')) %>%
  mutate(Hashtag = paste0('Hash', substr(Hashtag, 9, 10))) %>%
  column_to_rownames('Sample')
## Labels need to be a named list
HTO_labels <- setNames(rownames(HTO_table), HTO_table$Hashtag)
```

```{r}
objs <- lapply(objs, demuxAntibodyData, labels = HTO_labels, assay = 'ADT')
```

```{r}
hto_demux_results <- lapply(objs, function(x) table(x$HTO_classification))
```

```{r}
## Preserving doublets for now, just removing empty gems
objs <- lapply(objs, subset, subset = HTO_classification.global != 'Negative')
qc_receipts$remove_negative_gems <- lapply(objs, dim)
```

## Doublet detection scDblFinder

```{r}
objs <- lapply(objs, NormFindVarFeatScaleData)
```

```{r, eval = FALSE}
## Doublet detection requires super low-count cells to be removed. If you get
## errors saying 'Size factors should be positive', try removing low count cells .The
## vignette for scDblFinder suggests removing cells with less than 200-500 reads.
# objs <- lapply(objs, subset, subset = nCount_RNA > 200)
```

```{r}
objs <- lapply(objs, run_scDblFinder)
```

### Quick clustering check

```{r}
objs.tmp <- lapply(objs, function(x) {
  NormFindVarFeatScaleData(x) %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:20) %>%
    RunUMAP(obj, dims = 1:20)
})
umap.plots[['Doublets']] <- objs.tmp %>%
  lapply(., function(x) {
    x@meta.data <- x@meta.data %>% mutate(doublets = scDblFinder.class)
    x
    }) %>%
  plotUMAP_split(., group.by = "doublets", label.size=0, legend = T, reduction = 'GEX_UMAP')
umap.plots[['Doublets']] <- ggarrange(plotlist = umaps.rna[["doublets"]], ncol=3, nrow=2, 
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("doublets", face = "bold", size = 14))
```

### Remove doublets

```{r}
## For scDblFinder doublets
objs <- lapply(objs, function(x){
  subset(x, scDblFinder.class != 'doublet')
})
## For HTO doublets
objs <- lapply(objs, function(x){
  subset(x, HTO_classification.global != 'Doublet')
})
qc_receipts$doublet_removal <- lapply(objs, dim)
```

```{r}
doublet_counts <- data.frame(
  doublets = unlist(lapply(qc_receipts$remove_negative_gems, function(x) x[[2]])) -
    unlist(lapply(qc_receipts$doublet_removal, function(x) x[[2]])),
  total = unlist(lapply(qc_receipts$remove_negative_gems, function(x) x[[2]]))
) %>% 
  mutate(percent = round(100 * doublets / total, 2))
doublet_counts
```

## Removing low quality cells

This approach tests filtering cells based on cutoffs, outlier detection, or
both. Nothing is done to the data initially, so you can decide which filtering
regime to use. 


```{r}
objs <- lapply(objs,
              addQCmetrics,
              mito.pattern = "^MT", ## edit patterns as needed
              ribo.pattern = "^RP[SL]")
```

```{r}
## Define your thresholds. Helper function generate_cutoffs will 
## initialize a list using outlier detection. 
cutoffs <- lapply(objs, generate_capture_QC_cutoffs)
## you can always edit these directly
```

```{r, include=FALSE}
## this applies the cutoffs to your objects, and stores whether a cell 
## is filtered or not in the 'outliers' column
objs <- mapply(addQCfilter, objs, filterName = 'outliers', cutoffs)
```

```{r, message=TRUE, warning=FALSE, fig.width=18, fig.height=15}
## Generate dashboard plots of filtering regimes
cell_filtering_plots <- list()
## if you added multiple filters, say one with default settings and one with
## modified thresholds, you can plot all of them here. Specifying 'input'
## shows the unfiltered cells in a separate dashboard
for (filtName in c('input', 'outliers')) {
  cell_filtering_plots[[filtName]] <- plotQCRidgesJoint(objs, 
                                              filtName = filtName,
                                              title = filtName, 
                                              split_by = "capID", 
                                              color_by = "mitoRatio",
                                              cutoffs = cutoffs,
                                              facet_colors = TRUE)
}
cell_filtering_plots
```

```{r}
## choose a boolean metadata column to filter on
objs <- lapply(objs, subset, subset = outliers)
qc_receipts$qc_filtering <- lapply(objs, dim)
```

## Check cell cycle

```{r, include=FALSE}
## If human, cc genes were already loaded with Seurat's `cc.genes`

## If mouse data:
# source(here("R/get_cc_genes-Mus_musculus.R"))
# cc.genes <- list(s_genes, g2m_genes)
```

Score cell cycle phase
```{r, include=FALSE}
## ScoreCC also performs the normalization and scaling
objs <- lapply(objs, scoreCC, s.genes = cc.genes$s.genes, g2m.genes = cc.genes$g2m.genes)
```

Plot Phase on PCA
```{r, fig.width=30, fig.height=15}
lapply(objs, plotPCAs, features = "Phase") %>%
  unlist(recursive = FALSE) %>%
  ggarrange(plotlist = .)
```


We don't do any default correction for cell-cycle as it is a per-study consideration.
But the relevant metadata is stored with the object if you want to regress it out
downstream.

### Regress cell cycle

if you so choose

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=12}
umaps.rna[["Phase_pre_regression"]] <- objs %>%
  plotUMAP_split(., group.by = "Phase", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["Phase_pre_regression"]] <- ggarrange(plotlist = umaps.rna[["Phase_pre_regression"]], ncol=6, nrow=6, 
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("Phase", face = "bold", size = 14))
```

```{r}
regress_cc <- function(obj, ndim = 30){
  obj <- NormFindVarFeatScaleData(obj,
                           regress.out = c("S.Score", "G2M.Score"))
  obj <- RunPCA(obj)
  obj <- FindNeighbors(obj, dims = 1:ndim)
  obj <- RunUMAP(obj, dims = 1:ndim)
  obj
}
```

```{r}
objs.tmp <- lapply(objs, regress_cc)
```

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=12}
umaps.rna[["Phase_post_regression"]] <- objs.tmp %>%
  plotUMAP_split(., group.by = "Phase", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["Phase_post_regression"]] <- ggarrange(plotlist = umaps.rna[["Phase_post_regression"]],
                                               ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2], 
                                               common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("Phase", face = "bold", size = 14))
```

## Clustering

```{r}
## you could specify dimensions per capture, or you could just proceed
## with a single dim number for all captures.
## Still working out if it's necessary to be this selective
lapply(objs, ElbowPlot)
ndims <- 1:30
```


```{r}
## In MAPPLY format in case you wanted to provide different values for different 
## captures, but providing one argument to MoreArgs makes it the same for all 
## captures
objs <- mapply(FindNeighbors,
               objs,
               reduction = 'pca',
               MoreArgs = list(dims = ndims))
objs <- mapply(RunUMAP,
               objs,
               assay = 'RNA',
               reduction.name = 'GEX_UMAP',
               MoreArgs = list(dims = ndims))
objs <- lapply(objs, FindClusters)
```

## QC filtering summary

```{r}
## If you did hash demultiplexing, you may have additional bad gems from removing
## negatives or doublets identified by HTODemux
qc_filtering_summary <- as.data.frame(t(data.table::rbindlist(qc_receipts)[c(2,6,8),]))
colnames(qc_filtering_summary) <- c('Initial', 'qc_filtering', 'doublet_removal')
qc_filtering_summary <- qc_filtering_summary %>% 
  mutate(low_qual_cells = Initial - qc_filtering) %>%
  mutate(bad_gems = qc_filtering - doublet_removal) %>%
  dplyr::select(low_qual_cells, bad_gems, doublet_removal)

```

```{r, fig.width=8}
qc_filtering_results_plot <- reshape2::melt(as.matrix(qc_filtering_summary)) %>%
  ggplot(aes(x = Var1,
             y = value,
             group = Var2,
             fill = Var2)) + 
  geom_col() +
  theme_bw() + 
  labs(x = 'Sample', y = 'Cell count', fill = NULL) +
  scale_fill_manual(values = c('low_qual_cells' = 'darkred',
                               'bad_gems' = 'goldenrod',
                               # 'HTO_doublets' = 'orange',
                               'doublet_removal' = 'darkgreen'),
                    labels = c('Low quality cells',
                               'Bad GEMs',
                               # 'HTO Doublets',
                               'High quality cells'))
qc_filtering_results_plot
```

### Capture level metadata clustering

```{}
This hasn't been super robustly tested, so this might not work out of the box.
The idea here is to plot metadata such as phase, mito ratio, etc on the 
capture-level clusters to identify possible problematic groupings of cells
```


```{r}
umaps.rna[["clusters"]] <- objs %>%
  plotUMAP_split(.,
                 group.by = "seurat_clusters",
                 label.size = 0,
                 reduction = 'GEX_UMAP')
umaps.rna[["clusters"]] <- ggarrange(plotlist = umaps.rna[["clusters"]],
                                     ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("Clusters", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["Phase"]] <- objs %>%
  plotUMAP_split(., group.by = "Phase", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["Phase"]] <- ggarrange(plotlist = umaps.rna[["Phase"]],
                                  ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("Phase", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["nGene"]] <- objs %>%
  plotUMAP_split(., feature = "nGene", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["nGene"]] <- ggarrange(plotlist = umaps.rna[["nGene"]],
                                      ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("nGene", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["nUMI"]] <- objs %>%
  plotUMAP_split(., feature = "nUMI", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["nUMI"]] <- ggarrange(plotlist = umaps.rna[["nUMI"]],
                                       ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("nUMI", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["log10GenesPerUMI"]] <- objs %>%
  plotUMAP_split(., feature = "log10GenesPerUMI", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["log10GenesPerUMI"]] <- ggarrange(plotlist = umaps.rna[["log10GenesPerUMI"]],
                                             ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("log10GenesPerUMI", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["riboRatio"]] <- objs %>%
  plotUMAP_split(., feature = "riboRatio", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["riboRatio"]] <- ggarrange(plotlist = umaps.rna[["riboRatio"]],
                                      ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("riboRatio", face = "bold", size = 14))
```

```{r, fig.width=12, fig.height=8}
umaps.rna[["mitoRatio"]] <- objs %>%
  plotUMAP_split(., feature = "mitoRatio", label.size=0, legend = T, reduction = 'GEX_UMAP')

umaps.rna[["mitoRatio"]] <- ggarrange(plotlist = umaps.rna[["mitoRatio"]],
                                      ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],
          common.legend = T, legend = "right") %>%
  annotate_figure(top = ggpubr::text_grob("mitoRatio", face = "bold", size = 14))
```

# ADT Processing

This whole section has not super robustly tested

```{r}
objs <- lapply(objs,
               NormalizeData,
               assay = "ADT",
               normalization.method = "CLR", 
               margin = 2) ## Might use 1 for smaller panels

# cluster and run umap
objs <- lapply(objs, function(x) {
  DefaultAssay(x) <- "ADT"
  prots <- x@assays$ADT@counts %>% rownames()
  x <- x %>% 
    Seurat::FindNeighbors(dims = NULL,
                          features = prots,
                          k.param = 20,
                          verbose = FALSE) %>%
    Seurat::FindClusters(graph.name = "ADT_snn",
                         resolution = 0.8,
                         verbose = FALSE) %>%
    Seurat::RunUMAP(features = prots, 
                    verbose = FALSE,
                    reduction.name = 'ADT_UMAP')
})
```

## UMAPs of ADT expression
```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=18}
adt_umaps <- plotUMAP_split(objs,
                             group.by = "seurat_clusters",
                             label.size=0,
                             plot_smaller = T,
                             reduction = 'ADT_UMAP')

ggarrange(plotlist = adt_umaps, ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],) 

```

## Heatmaps comparing clusters to ADT expression

```{r}
heatmapClusterADT <- function(x,
                              title = x@project.name, 
                              silent = TRUE) {
  prots <- x@assays$ADT@counts %>% rownames()
  adt_plot <- cbind(x@meta.data, as.data.frame(t(x@assays$ADT@data))) %>%
    dplyr::group_by(seurat_clusters) %>%
    dplyr::summarize_at(.vars = prots, .funs = median) %>%
    tibble::remove_rownames() %>%
    tibble::column_to_rownames("seurat_clusters") 
  
  pheatmap::pheatmap(t(adt_plot),
                     color = viridis::viridis(25, option = "B"),
                     fonstize_row = 8, border_color = NA, silent = silent,
                     legend = FALSE, treeheight_row = 10, treeheight_col = 10,
                     fontsize_row = 4, fontsize_col = 6, main = title) %>% 
    ggplotify::as.ggplot()
}
```


```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=20}

# adt_heatmaps <- lapply(objs_filt_mapped[-7], function(x) {
#   heatmapClusterADT(x, title = '')
# })

adt_heatmaps <- lapply(
  seq_along(objs),
  function(i, x=objs) {
    heatmapClusterADT(objs[[i]], title = names(objs)[[i]])
})
```


```{r, fig.width=12, fig.height=20}
#adt_heatmaps
ggarrange(plotlist=adt_heatmaps, ncol=umaps.rna$dims[1], nrow=umaps.rna$dims[2],) %>%
  annotate_figure(top = ggpubr::text_grob("", face = "bold", size = 14))
```

# Integration

```{r}
obj_post_harmony_int <- runIntegration(objs,
                                       integration_method = 'harmony',
                                       dim.reduct = "rpca",
                                       harmony.group.by.vars = c('capID', 'hash.ID'))
```

## Cluster membership plots

```{r}
DimPlot(obj_post_harmony_int, group.by = 'capID', raster = FALSE)
```

```{r}
DimPlot(obj_post_harmony_int, split.by = 'capID', raster = FALSE, ncol = 5)
```

## Cluster membership tables

Visuals of how well integrated the captures were

```{r}
integration_table <- table(obj_post_harmony_int$capID, obj_post_harmony_int$seurat_clusters) %>%
  as.data.frame() %>%
  reshape2::dcast(Var1 ~ Var2) %>%
  column_to_rownames('Var1') %>%
  as.matrix()
```

```{r}
cap_proportions_DT <-  DT::datatable(
  round(100*proportions(integration_table, margin = 1), 1),
  options = list(pageLength = 15,
                 searching=FALSE,
                 ordering=FALSE,
                 lengthChange = FALSE)) %>%
  DT::formatStyle(colnames(integration_table),
                  background = DT::styleColorBar(c(0,100), 'lightblue'))
cap_proportions_DT
```

```{r}
clus_proportions_DT <-  DT::datatable(
  round(100*proportions(integration_table, margin = 2), 1),
  options = list(pageLength = 15,
                 searching=FALSE,
                 ordering=FALSE,
                 lengthChange = FALSE)) %>%
  DT::formatStyle(colnames(integration_table),
                  background = DT::styleColorBar(c(0,100), 'lightblue'))
clus_proportions_DT
```

```{r}
## Moving cell annotation to the next script, so this function may need to be tweaked
# plot_list <- list()
# plot_list <- plotIntegrationDiagnostics(plot_list = plot_list,
#                                         seurat.obj = obj_post_harmony_int,
#                                         sample_col = 'capID',
#                                         subset_id = 'full',
#                                         integration_name = 'Harmony',
#                                         cell_labels = 'predicted.cell_type_lvl1',
#                                         res = c(0.9),
#                                         pt.size = .1)
# plot_list$full$Harmony$predicted.cell_type_lvl2 <- plotUmapIntegrated(obj_post_harmony_int, group.by = 'predicted.cell_type_lvl2', label.size = 3, pt.size = .1, title = 'predicted.cell_type_lvl2')
```

# SaveRDS

```{r}
saveRDS(obj_post_harmony_int,here('saved_rds/obj_post_harmony.Rds'), compress=FALSE)
obj_post_harmony_int <- readRDS(here('saved_rds/obj_post_harmony.Rds'))
# rm(objs)
gc()
```

# Render report

```{r}
## This only works if all the variables called in the format file 
## are in the current environment. Check the format file to tweak
## aesthetics 
rmarkdown::render(here('analysis_scripts/02_processing_template.format.Rmd'),
                  output_file = '02_QC.html',
                  output_dir = here('reports'))
```

