---
title: "scRNAseq celltype ID template"
author: "ENPRC Genomics Core"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(Seurat)
library(tidyverse)
library(here)
library(gencoreSC)
# library(RColorBrewer)
library(clustree)
library(ggpubr)
library(DT)
knitr::opts_chunk$set(echo=FALSE, warning=TRUE, message=TRUE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE,
                      fig.align = 'center')#, include=FALSE)
```


# Load data

```{r}
processed_obj <- readRDS(here('saved_rds/obj_final_post_processing.Rds'))
```

## Reference mapping

```{}
Somewhat TBD if this belongs in the processing step, it's long but pretty easy
to run this way so I'd like to have it done prior to the 'analysis' so that
object is stable. 
```



# SingleR

```{r}
## Load a reference dataset in
ln_ref_data <- SeuratDisk::LoadH5Seurat('/Volumes/genomelab/illumina/runs/scrna_seq_ref/tabula_sapiens/ts_lymph_downsampled_copy.h5Seurat')

## SingleR can work on merged object, or lapply over captures
# objs_filt_mapped <- merge(objs_filt_mapped[[1]], objs_filt_mapped[2:length(objs_filt_mapped)])

c(obj_filt_mapped, singler_out) %<-% lapply(objs_filt, runSingleR,
                          ref = ln_ref_data@assays$RNA@data,
                          labels = ln_ref_data$labels,
                          meta.prefix = 'singleR.out',
                          de.method = 'wilcox')

```

## Plotting mapping results

```{r, fig.width=14}
singleR_heatmap <- plotScoreHeatmap(singleR_results$Cap1)
singleR_heatmap
```

## Reference prediction markers

```{r, fig.width = 17, fig.height = 12}
# This could be a nice thing to add to the package if I'm able to ggarrange it
plotSingleRMarkerHeatmaps <- function(seurat_obj, ref_markers, labels, show_n_markers = 20) {
  # Plotting aesthetics are much better with a shorter legend title
  seurat_obj@meta.data <- seurat_obj@meta.data %>%
    mutate(labels = .data[[labels]])
  
  # Convert seurat obj to sce for input to scater::plotHeatmap
  obj_sce <- seurat_obj %>% 
    Seurat::as.SingleCellExperiment(assay = "RNA") %>% 
    scater::logNormCounts()
  
  # Find markers from the empirical data to compare to the reference based markers in `ref_markers`
  empirical.markers <- obj_sce %>%
    scran::findMarkers(., obj_sce[["labels"]], direction="up")
  
  # Plot a separate heatmap for each of the marker sets applied
  heatmaps <- lapply(names(ref_markers), function(x) {
    ref_markers_celltype <- unique(unlist(ref_markers[[x]]))
    m <- match(ref_markers_celltype, rownames(empirical.markers[[x]]))
    m <- ref_markers_celltype[rank(m) <= show_n_markers]
    
    heatmap <- scater::plotHeatmap(obj_sce, order_columns_by = "labels", features = m, 
                           silent = TRUE, main = x)[[4]]
    return(heatmap)
  })
  names(heatmaps) <- names(ref_markers)
  return(heatmaps)
}

# Get markers for each annotation label
singlerMarkers_all <-  metadata(singler_out)$de.genes

singlerMarker_heatmaps <- plotSingleRMarkerHeatmaps(seurat_obj = processed_obj, 
                                                    ref_markers = singlerMarkers_all, 
                                                    labels = "singleR.pruned.labels")
```


```{r, fig.width = 23, fig.height = 16}
ggarrange(plotlist = singlerMarker_heatmaps,
          ncol = 3, nrow = 3)
```

# Selecting resolution

```{r}
clustree_obj <- processed_obj
for (res in seq(0.2, 1.5, .1)){
  ## in case you rerun this with different ranges of res
  if (!(paste0('RNA_snn_res.', res) %in% colnames(clustree_obj@meta.data))){
    clustree_data <- FindClusters(processed_obj, resolution = res) 
  } else{
    print(paste0('Resolution ', res, ' already calculated, skipping to next res'))
  }
}
```

```{r, fig.height=14, fig.width=14}
cluster_tree <- clustree(clustree_data)
cluster_tree
```

```{r}
## See how a resolution looks in UMAP space
DimPlot(clustree_obj, group.by = 'RNA_snn_res.1.3')
```

```{r}
## Chose a resolution to move forward with
processed_obj <- FindClusters(processed_obj, resolution = )
DimPlot(processed_obj, pt.size = 0.5)
```

# Cluster plots

## Dashboard

```{r, fig.width=10, fig.height=6}
## if you want to plot some metadata with your new clusters
clustering_dashboard_plot <- ggarrange(
  DimPlot(processed_obj, group.by = 'hash.ID'),
  DimPlot(processed_obj, group.by = 'seurat_clusters'),
  DimPlot(processed_obj, group.by = 'Phase'),
  DimPlot(processed_obj, group.by = 'singleR.pruned.labels', label = FALSE),
  nrow = 2, ncol = 2)
clustering_dashboard_plot
```

# Reference annotation predictions

```{r}
cluster_label_table <- 
  processed_obj@meta.data %>% 
  group_by(seurat_clusters) %>%
  count(singleR.pruned.labels) %>%
  mutate(pct = round(n / sum(n) * 100, 2)) %>%
  slice_max(order_by = pct, n = 1) %>%
  select(-n) 
cluster_label_table
```

## Prediction facetplot

```{r, fig.height=12, fig.width=10, warning = FALSE}
hpca_facet_plot <- plotRefMapScoresFacet(processed_obj,
                      label_column = 'singleR.out.pruned.labels', 
                      label_score_column = 'singleR.out.delta.next',
                      clusters_column = 'cluster_labels')
hpca_facet_plot
```

## Surface markers

```{r, fig.height=8}
heatmaps.adt <- featureHeatmapByCluster(processed_obj, title = 'ADT Data by cluster', assay = 'ADT')
heatmaps.adt
```

```{r}
## saving the clustered obj lets you call findMarkers as a background script
## see the script at R/findMarkers.R
saveRDS(processed_obj, file = here('saved_rds/obj_post-initial-clustering.Rds'), compress = FALSE)
```

# cluster markers

```{r}
## Found using the saved rds and the background job script R/findMarkers.R
cluster_markers <- readRDS(here('saved_rds/cluster_markers_test.Rds'))
```

```{r}
cluster_marker_DT <- cluster_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val, p_val_adj, pct.1, pct.2) %>%
  filter(p_val_adj <= 0.05) %>%
  mutate(avg_log2FC = round(avg_log2FC, 3),
         p_val = format(p_val, scientific = TRUE, digits = 3),
         p_val_adj = format(p_val_adj, scientific = TRUE, digits = 3),
         pct.1 = round(pct.1, 3)*100,
         pct.2 = round(pct.2, 3)*100) %>%
  datatable(
    rownames = FALSE,
    colnames = c('Cluster',
      'Gene', 
      'Avg. Log2 fold change',
      'p-value', 
      'Adj. p-value',
      '% expressed in cluster',
      '% expressed outside cluster'), 
    caption='Markers within cluster relative to outside cluster',
    filter='top',
    autoHideNavigation = TRUE,
    extensions = c('Buttons', 'Scroller'),
    options = list(
      dom = 'Bfrtip',
      buttons = c('copy', 'csv'),
      deferRender = TRUE
    )) #%>%
    # formatRound(c('avg_log2FC'), digits=3) %>%
    # formatRound(c('pct.1', 'pct.2'), digits = 2) %>%
    # formatSignif(c('p_val', 'p_val_adj'))
cluster_marker_DT
```

# Save object

```{r}
saveRDS(processed_obj, here('saved_rds/obj_IDd_clusters.Rds'))
```

# Render report

```{r}
rmarkdown::render(here("analysis_scripts/03-clustering.format.Rmd"),
                  output_file = '03-clustering.html',
                  output_dir = here('reports'))
```
