---
title: "scRNAseq celltype ID template"
author: "ENPRC Genomics Core"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(Seurat)
library(tidyverse)
library(here)
library(gencoreSC)
# library(RColorBrewer)
library(clustree)
library(UCell)
library(ggpubr)
library(DT)
knitr::opts_chunk$set(echo=FALSE, warning=TRUE, message=TRUE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE,
                      fig.align = 'center')#, include=FALSE)
```


# Load data

```{r}
obj <- readRDS(here('saved_rds/obj_post_harmony.Rds'))
```

# Reference mapping

```{r}
## Load a reference dataset in
ref <- SeuratDisk::LoadH5Seurat('/Volumes/genomelab/illumina/runs/scrna_seq_ref/tabula_sapiens/ts_lymph_downsampled_copy.h5Seurat')
## or use celldex
#ref <- celldex::MonacoImmuneData()

## SingleR can work on merged object, or lapply over captures
# objs_filt_mapped <- merge(objs_filt_mapped[[1]], objs_filt_mapped[2:length(objs_filt_mapped)])

c(obj, singleR_results) %<-% runSingleR(obj, 
                                        ref = ref@assays$RNA@data,
                                        labels = ref$labels,
                                        meta.prefix = 'singleR.out',
                                        de.method = 'wilcox')

```

## Plotting mapping results

```{r, fig.width=14}
singleR_heatmap <- plotScoreHeatmap(singleR_results$Cap1)
singleR_heatmap
```

## Reference prediction markers

```{r, fig.width = 17, fig.height = 12}
# This could be a nice thing to add to the package if I'm able to ggarrange it
plotSingleRMarkerHeatmaps <- function(seurat_obj, ref_markers, labels, show_n_markers = 20) {
  # Plotting aesthetics are much better with a shorter legend title
  seurat_obj@meta.data <- seurat_obj@meta.data %>%
    mutate(labels = .data[[labels]])
  
  # Convert seurat obj to sce for input to scater::plotHeatmap
  obj_sce <- seurat_obj %>% 
    Seurat::as.SingleCellExperiment(assay = "RNA") %>% 
    scater::logNormCounts()
  
  # Find markers from the empirical data to compare to the reference based markers in `ref_markers`
  empirical.markers <- obj_sce %>%
    scran::findMarkers(., obj_sce[["labels"]], direction="up")
  
  # Plot a separate heatmap for each of the marker sets applied
  heatmaps <- lapply(names(ref_markers), function(x) {
    ref_markers_celltype <- unique(unlist(ref_markers[[x]]))
    m <- match(ref_markers_celltype, rownames(empirical.markers[[x]]))
    m <- ref_markers_celltype[rank(m) <= show_n_markers]
    
    heatmap <- scater::plotHeatmap(obj_sce, order_columns_by = "labels", features = m, 
                           silent = TRUE, main = x)[[4]]
    return(heatmap)
  })
  names(heatmaps) <- names(ref_markers)
  return(heatmaps)
}

# Get markers for each annotation label
singlerMarkers_all <-  metadata(singler_out)$de.genes

singlerMarker_heatmaps <- plotSingleRMarkerHeatmaps(seurat_obj = obj, 
                                                    ref_markers = singlerMarkers_all, 
                                                    labels = "singleR.pruned.labels")
```


```{r, fig.width = 23, fig.height = 16}
ggarrange(plotlist = singlerMarker_heatmaps,
          ncol = 3, nrow = 3)
```

# Selecting resolution

```{r}
clustree_obj <- obj
for (res in seq(0.2, 1.5, .1)){
  ## in case you rerun this with different ranges of res
  if (!(paste0('RNA_snn_res.', res) %in% colnames(clustree_obj@meta.data))){
    clustree_data <- FindClusters(processed_obj, resolution = res) 
  } else{
    print(paste0('Resolution ', res, ' already calculated, skipping to next res'))
  }
}
```

```{r, fig.height=14, fig.width=14}
cluster_tree <- clustree(clustree_data)
cluster_tree
```

```{r}
## See how a resolution looks in UMAP space
DimPlot(clustree_obj, group.by = 'RNA_snn_res.1.3')
```

```{r}
## Chose a resolution to move forward with
obj <- FindClusters(obj, resolution = )
DimPlot(obj, pt.size = 0.5)
```

# Cluster plots

## Dashboard

```{r, fig.width=10, fig.height=6}
## if you want to plot some metadata with your new clusters
clustering_dashboard_plot <- ggarrange(
  DimPlot(obj, group.by = 'hash.ID'),
  DimPlot(obj, group.by = 'seurat_clusters'),
  # DimPlot(obj, group.by = 'Phase'),
  DimPlot(obj, group.by = 'singleR.pruned.labels', label = FALSE),
  nrow = 2, ncol = 2)
clustering_dashboard_plot
```

## Reference predictions by cluster

```{r}
cluster_label_table <- 
  obj@meta.data %>% 
  group_by(seurat_clusters) %>%
  count(singleR.pruned.labels) %>%
  mutate(pct = round(n / sum(n) * 100, 2)) %>%
  slice_max(order_by = pct, n = 1) %>%
  select(-n) 
cluster_label_table
```

### Prediction facetplot

```{r, fig.height=12, fig.width=10, warning = FALSE}
ref_facet_plot <- plotRefMapScoresFacet(obj,
                      label_column = 'singleR.out.pruned.labels', 
                      label_score_column = 'singleR.out.delta.next',
                      clusters_column = 'cluster_labels')
ref_facet_plot
```

## Surface markers

```{r, fig.height=8}
heatmaps.adt <- gencoreSC::featureHeatmapByCluster(obj, title = 'ADT Data by cluster', assay = 'ADT')
heatmaps.adt
```

```{r}
## saving the clustered obj lets you call findMarkers as a background script
## see the script at R/findMarkers.R
saveRDS(obj, file = here('saved_rds/obj_post-initial-clustering.Rds'), compress = FALSE)
```

# cluster markers

```{r}
## Found using the saved rds and the background job script R/findMarkers.R
cluster_markers <- readRDS(here('saved_rds/cluster_markers_test.Rds'))
```

```{r}
cluster_marker_DT <- cluster_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val, p_val_adj, pct.1, pct.2) %>%
  filter(p_val_adj <= 0.05) %>%
  mutate(avg_log2FC = round(avg_log2FC, 3),
         p_val = format(p_val, scientific = TRUE, digits = 3),
         p_val_adj = format(p_val_adj, scientific = TRUE, digits = 3),
         pct.1 = round(pct.1, 3)*100,
         pct.2 = round(pct.2, 3)*100) %>%
  datatable(
    rownames = FALSE,
    colnames = c('Cluster',
      'Gene', 
      'Avg. Log2 fold change',
      'p-value', 
      'Adj. p-value',
      '% expressed in cluster',
      '% expressed outside cluster'), 
    caption='Markers within cluster relative to outside cluster',
    filter='top',
    autoHideNavigation = TRUE,
    extensions = c('Buttons', 'Scroller'),
    options = list(
      dom = 'Bfrtip',
      buttons = c('copy', 'csv'),
      deferRender = TRUE
    )) #%>%
    # formatRound(c('avg_log2FC'), digits=3) %>%
    # formatRound(c('pct.1', 'pct.2'), digits = 2) %>%
    # formatSignif(c('p_val', 'p_val_adj'))
cluster_marker_DT
```

# Known markers

```{r}
known_markers <- openxlsx::read.xlsx(here('../Notes/Innate Cell Markers.xlsx'))
known_markers <- unstack(select(known_markers, c(Mmul10.gene.symbol, Celltype)))
known_markers
```

```{r}
known_marker_dotplots <- lapply(names(known_markers),
                                   function(x) DotPlot(obj, features = known_markers[x]) + 
                                     theme(axis.text.x = element_text(angle=30, hjust = 1)))
names(known_marker_dotplots) <- names(known_markers)
known_marker_dotplots
```

## Module scores

```{r}
for (module in names(known_markers)){
  obj <- UCell::AddModuleScore_UCell(obj, feature = known_markers[module])
}
```

```{r, fig.height=9}
FeaturePlot(obj, features = paste0(names(known_markers), '_UCell'))
```

```{r, fig.width=12}
VlnPlot(obj, paste0(names(known_markers), '_UCell'), pt.size = 0, y.max = 1, same.y.lims = TRUE)
```

# Assigning celltypes

```{r}
cluster_identities <- c(
  '0. ', #0
  '1. ', 
  '2. ', 
  '3. ',    
  '4. ',    
  '5. ',      #5
  '6. ',        
  '7. ', 
  '8. ',    
  '9. ',    
  '10. ', #10
  '11. ',
  '12. ',
  '13. ', 
  '14. ', 
  '15. ', #15
  '16. ',
  '17. '
)
obj$cluster_labels <- plyr::mapvalues(
  obj$seurat_clusters, 
  from = levels(obj$seurat_clusters),
  to = cluster_identities
)
```

```{r}
DimPlot(obj, group.by = 'cluster_labels')
```

## Cluster membership tables

These will need to be modified/generalized for your metadata variables of interest

```{r}
cluster_membership_table <- as.data.frame(table(obj$capID, obj$cluster_labels)) %>%
  reshape2::dcast(Var1 ~ Var2, value.var = 'Freq') %>%
cluster_membership_table
```

This can show proportions by row or column in a visual table to quickly
check for imbalances

```{r}
cluster_membership_table_2 <- table(obj$previous_labels,
                           obj$seurat_clusters) %>% 
  proportions(margin = 2) 
cluster_membership_table_2 <- round(cluster_membership_table_2 * 100, 1) %>%
  as.data.frame.matrix()

cluster_membership_table_2 <-  datatable(
  cluster_membership_table_2,
  options = list(pageLength = 45,
                 searching=FALSE,
                 ordering=FALSE,
                 lengthChange = FALSE,
                 dom = 'Bfrtip',
                 buttons = c('copy', 'csv'),
                 deferRender = TRUE)) %>%
  DT::formatStyle(names(cluster_membership_table_2),
                  background = styleColorBar(c(0,100), 'lightblue'))
cluster_membership_table_2
```


# Save object

```{r}
saveRDS(obj, here('saved_rds/obj_post_clustering.Rds'))
```

# Render report

```{r}
rmarkdown::render(here("analysis_scripts/03_clustering&cluster-labeling.format.Rmd"),
                  output_file = '03_clustering&cluster-labeling.html',
                  output_dir = here('reports'))
```
