---
title: 'p22091_Peng_analysis'
author: "Emory Gencore"
date: '`r Sys.Date()`'
output: 
  rmdformats::robobook:
    fig_width: 8
    gallery: true
    thumbnails: false
    toc_depth: 1
---
    
<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r, include=FALSE}
library(ggrepel)
library(DT)
library(Seurat)
library(tidyverse)
library(reshape2)
library(dplyr)
library(biomaRt)
library(fgsea)
#library(org.Mm.eg.db)
#library(clusterProfiler)
library(ape)
library(MAST)
library(DoubletFinder)
library(Azimuth)
library(SeuratData)
library(patchwork)
#library(package = affyLib, )
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, cache=TRUE, cache.lazy=FALSE)
config <- yaml::yaml.load_file("scRNA_GEX_config.yml")
source('scRNAseq_functions.R')
```


```{r}
combined.obj <- readRDS('saved_rds/combined_obj_postQC.Rds')
DefaultAssay(combined.obj) <- "integrated"
```

# Overview

We are collecting FNA samples from rhesus macaques pre and post chimeric influenza vaccination. We aim to identify the subsets of dendritic cells and monocytes being trafficked to the lymph node post cH5 mRNA immunization.

```{r}
DimPlot(combined.obj, reduction = 'umap', label=TRUE)
```

# Reference mapping

## Tabula sapiens map query

```{r reference_data, include=FALSE,}
## Subsetting tabula sapiens
# ref <- readRDS("/Volumes/yerkes/genomelab/illumina/runs/Genome_references/tabula_sapiens/local.rds")
# ref@assays$RNA@key <- "rna_"
# ref <- subset(x=ref, subset = tissue == "lymph node")
# saveRDS(ref, file='tabula_sapiens_lymph_subset.Rds')
ref <- readRDS('tabula_sapiens_lymph_subset.Rds')
```

```{r, include=FALSE, cache=TRUE}
# library(biomaRt)
# human <- useMart('ensembl', dataset = 'hsapiens_gene_ensembl')
# rhesus <- useMart('ensembl', dataset = 'mmulatta_gene_ensembl')
# annot_table <- getLDS(
#   mart = human,
#   attributes = c('ensembl_gene_id','external_gene_name','chromosome_name'),
#   martL = rhesus,
#   attributesL = c('ensembl_gene_id','external_gene_name','chromosome_name','gene_biotype'))

## Reformatting reference for macaque genes
transfer_table <- read.csv('human_to_macaque_table.txt', sep='\t')
human_to_macaque_symbols <- function(gene, transfer_table){
  if (gene %in% transfer_table$Human.gene.name){
    transfer_table = transfer_table %>% filter(Human.gene.name == gene)
    
  } else if (gene %in% transfer_table$Human.gene.stable.ID){
    transfer_table = transfer_table %>% filter(Human.gene.stable.ID == gene)
  } else {
    return('NA')
  }
  if (!is.null(transfer_table$Gene.name )) {
    return(transfer_table[1,]$Gene.name)
  } else  if (!is.null(transfer_table$Gene.stable.ID)) {
    return(transfer_table[1,]$Gene.stable.ID)
  } else {
    return('NA')
  }
}
fixed_gene_names <- unlist(lapply(rownames(ref),
                                  human_to_macaque_symbols,
                                  transfer_table = transfer_table))
fixed_gene_names[nchar(fixed_gene_names)==0] <- "NA"
counts <- GetAssayData(ref,assay = "RNA",slot = "counts")
rownames(counts) <- fixed_gene_names
counts <- counts[!rownames(counts)=="NA",]
ref <- CreateSeuratObject(counts = counts[!duplicated(rownames(counts)),], meta.data = ref@meta.data)

## Make the umap for the reference ahead of time
ref <- SCTransform(ref)
ref <- RunPCA(ref)
ref <- RunUMAP(ref, dims = 1:30, reduction = 'pca', return.model = TRUE)
```


```{r, include=FALSE}
combined.obj <- seurat_map_reference(combined.obj, ref=ref)
```

```{r genes_to_exclude, include=FALSE, eval=FALSE}
# 
# GFF <- read.gff(file.path(config$rootDir, "/Genome_references/mus_musculus/GRCm38/ensembl_102/Mus_musculus.GRCm38.102.gtf"))
# 
# 
# GFF$gene_biotype <- unlist(lapply(GFF$attributes, extract_attributes, "gene_biotype"))
# GFF$name <- unlist(lapply(GFF$attributes, extract_attributes, "name"))
# coding.genes <- unique(GFF %>% filter(gene_biotype=="protein_coding") %>% select(name))
# ig.genes <- unique(GFF %>% filter(grepl("IG_", gene_biotype)) %>% select(name))
# tr.genes <- unique(GFF %>% filter(grepl("TR_", gene_biotype)) %>% select(name))
# mt.genes <- unique(GFF %>% filter(grepl("mt-", name)) %>% select(name))
# #genes <- convertMouseGeneList(humGenes)

```

```{r, fig.width=8, fig.height=6}
### Showing how query maps to ref
p1 <- DimPlot(ref, reduction = "umap", group.by = "cell_type",
              label = TRUE, label.size = 3, repel = TRUE) +
  NoLegend() + ggtitle("Reference annotations")
p2 <- DimPlot(combined.obj, reduction = "ref.umap",
              group.by = "predicted.cell_type", label = TRUE, 
              label.size = 3, repel = TRUE) + 
  NoLegend() + ggtitle("Query transferred labels")
p1 + p2
```

```{r, fig.height=6, fig.width=8, warning=FALSE}
## Can specify facet order to rearrange facets, otherwise it will just be sequential
map_prediction_facetplot(combined.obj,
                         label_column = 'predicted.cell_type', 
                         score_column = 'predicted.cell_type.score',
                         min_freq = 0.05,
                         facet_order = c(3,9,1,12,13,15,14,0,10,4,2,16,7,5,6,8,11) )
```

```{r}
mapquery_top_results(combined.obj,
                     label_column = 'predicted.cell_type', 
                     score_column = 'predicted.cell_type.score',
                     top_n = 2) %>%
  knitr::kable(digits = 3, 
               col.names = c('Cluster',
                             'Predicted cell type',
                             'Median confidence score', 
                             'Proportion of cluster'),
               caption = 'Top 2 most common calls for each cluster')
```

## Azimuth PBMCs map query 

```{r, include=FALSE}
# library(pbmcref.SeuratData)
# InstallData('pbmc3k')
# LoadData('pbmcref')
pcmb_ref <- LoadData('pbmcref', "azimuth")
combined.obj.pbmc <- RunAzimuth(combined.obj, reference = 'pbmcref', umap.name='azimuth.umap')
```

```{r, fig.width=8, fig.height=6}
p1 <- DimPlot(pcmb_ref$map, reduction = "refUMAP", group.by = "celltype.l2",
              label = TRUE, label.size = 3, repel = TRUE) +
  NoLegend() + ggtitle("Reference annotations")
p2 <- DimPlot(combined.obj.pbmc, reduction = "azimuth.umap",
              group.by = "predicted.celltype.l2", label = TRUE, 
              label.size = 3, repel = TRUE) + 
  NoLegend() + ggtitle("Query transferred labels")
p1 + p2
```

```{r, fig.height=6, fig.width=8, warning=FALSE}
map_prediction_facetplot(combined.obj.pbmc,
                         label_column = 'predicted.celltype.l2', 
                         score_column = 'predicted.celltype.l2.score',
                         min_freq = 0.05)

```

```{r}
mapquery_top_results(combined.obj.pbmc,
                     label_column = 'predicted.celltype.l2', 
                     score_column = 'predicted.celltype.l2.score',
                     top_n = 2) %>%
  knitr::kable(digits = 3, 
               col.names = c('Cluster',
                             'Predicted cell type',
                             'Median confidence score', 
                             'Proportion of cluster'),
               caption = 'Top 2 most common calls for each cluster')
```

***

# Cluster Markers

We compare cells within a cluster to all other clusters to calculate differential gene expression between clusters. This calculation considers expression variability and detection variability to determine significance. 

```{r find_all_markers, include=FALSE}
if (preran){
  clusters.markers <- readRDS("clusters_markers.Rds")
} else {
  clusters.markers <- data.frame()
  for (cluster in as.numeric(levels(combined.obj$seurat_clusters))){
    markers <- FindConservedMarkers(combined.obj,
                                    ident.1 = cluster,
                                    grouping.var = 'orig.ident',
                                    assay = 'RNA',
                                    min.pct = 0.25,
                                    logfc.threshold = log2(1.5),
                                    only.pos = TRUE)
    if (nrow(markers) > 0 ) {
      markers$cluster <- cluster
      clusters.markers <- rbind(clusters.markers, markers)
    }
  }
  # FindAllMarkers(combined.obj, min.pct = 0.25, 
  #                                    logfc.threshold = log2(1.5),
  #                                    only.pos = TRUE))
clusters.markers <- clusters.markers %>% rownames_to_column('gene')
saveRDS(clusters.markers, "saved_rds/clusters_markers.Rds")
}
```
```{r featureplot_top_markers, fig.width=10, fig.height=12}
# topmarkers <- clusters.markers %>%
#     group_by(cluster) %>%
#     top_n(n = 1, wt = avg_log2FC)
#FeaturePlot(combined.obj, features=topmarkers$gene)
```

<!-- Heatmap of the top 5 gene identifiers for each cluster. -->

```{r heatmap_top_markers, fig.height=16, fig.width=16}
# combined.obj <- ScaleData(combined.obj, features = rownames(combined.obj@assays$RNA), assay = 'RNA')
# top5 <- clusters.markers %>%
#     group_by(cluster) %>%
#     top_n(n = 5, wt = minimump_p_val) 
# DoHeatmap(combined.obj, features = top5$gene, assay='RNA')
```

Table of all markers for all clusters. 

```{r datatable_markers}
# DT::datatable(clusters.markers %>% mutate_if(is.numeric, ~round(., 4) %>% select(cluster, gene, avg_log2FC, p_val, p_val_adj, pct.1, pct.2)), filter = 'top', rownames = FALSE, colnames=c("Cluster", "Gene", "Mean log2 fold change", "P value", "Adj. p val", "% cluster", "% outside cluster"))
DT::datatable(clusters.markers %>% mutate_if(is.numeric, ~round(., 4)) %>% dplyr::relocate(cluster, gene, ends_with('log2FC')) %>% dplyr::select(-ends_with('p_val')), filter = 'top', rownames = FALSE)
```

***

# Tracking known cell-type markers

```{r}
DotPlot(combined.obj, features=c('PECAM1', 'JAM2', 'PROX1', 'ACTA2', 'CCL19', 'CCL21','CR2', 'DCN', 'SDC1', 'CCR7', 'CD83'), assay = 'RNA')
```

***

# Final cluster labels

```{r label_clusters, fig.width=10, fig.height=10}
cluster_tags <- c(
  "0. NT/CM CD8+ T cells",    #0
  "1. NT/CM CD4+ T cells",
  "2. EM CD4+ T cells",
  "3. Neutrophils",
  "4. EM CD8+ T cells",#Neutrophils",    
  "5. Monocytes/Macrophages",               #5
  "6. Plasmacytoid DC?",
  "7. Neutrophils",
  "8. DC (Myeloid?)",
  "9. Basophil or Erythroid",          
  "10. Neutrophil/Eosinophil?",                   #10
  "11. Monocytes",
  "12. NK cells",
  "13. DC"
  )
combined.obj$labels <- "NA"
combined.obj$labels <- plyr::mapvalues(combined.obj$seurat_clusters, 
                                       from=0:(length(unique(combined.obj$seurat_clusters))-1), to=cluster_tags)
DimPlot(combined.obj, group.by = "labels", label = TRUE, label.box=TRUE, repel=TRUE) + theme(legend.position="none") + ggtitle("Manual labels")
```

# Differential gene expression {.tabset}

DGE within clusters of interest. Each comparison has a tab with the following components:

* Number of cells from each condition
* Conserved markers for the cluster are shown as safety check to confirm selection and annotation (these are the same as the previous section, just specific to the cluster of interest)
* Differentially expressed genes tables


DGE tables have the following values:

* Adj. p val : Adjusted p-value, based on Bonferroni correction using all features in the dataset.
* Mean log2 fold change : log fold-change of the average expression between the two groups. **Positive values indicate that the feature is more highly expressed in the  LysM-cre group**
* pct.1 : The percentage of cells where the feature is detected in the first group
* pct.2 : The percentage of cells where the feature is detected in the second group
    
    
```{r}
per_cluster_DGE <- function(cluster_num){
  #cat('Number of cells from each condition')
  counts <- table((combined.obj@meta.data %>% filter(seurat_clusters == cluster_num))$orig.ident)
  
  #cat('Conserved cluster markers')
  consmarkers <- datatable(conserved_markers_pos %>% filter(cluster == cluster_num) %>% mutate_if(is.numeric, ~round(., 2)) %>% filter(max_pval<0.05) %>% dplyr::select(cluster, Gene, avg_log2FC, max_pval, `LysM-cre_pct.1`, `LysM-cre_pct.2`, `Ptpn11-E76K-pos-LysM-cre_pct.1`, `Ptpn11-E76K-pos-LysM-cre_pct.2`), filter = 'top', rownames = FALSE, colnames=c("Cluster", "Gene", "Mean log2 fold-change across conditions in cluster", "Max Adj. p val", "% detected in\nLysm-cre in cluster", "% detected in Lysm-cre noutside cluster", 
                              "% detected in Ptpn11-E76K-pos in cluster", "% detected in Ptpn11-E76K-pos outside cluster"              ), caption = 'Marker genes for cluster')
  
  markers <- FindMarkers(subset(combined.obj, subset = seurat_clusters == cluster_num), group.by = 'orig.ident', ident.1 = "LysM-cre", ident.2 = "Ptpn11-E76K-pos-LysM-cre", assay = 'RNA', min.pct = 0.25, verbose = FALSE, logfc.threshold = log2(1.5), test.use = 'MAST') %>% rownames_to_column('Gene')
  
  #cat('DGE')
  
  dge <- datatable(markers %>% mutate_if(is.numeric, ~round(., 4)) %>% filter(p_val_adj<0.05) %>% dplyr::select(Gene, avg_log2FC, p_val_adj, pct.1, pct.2), filter = 'top', rownames = FALSE, colnames=c("Gene", "Mean log2\nfold change", "Adj. p val", "% detected\nin LysM-cre", "% detected\nin Ptpn11-E76K-pos-LysM-cre"), caption = 'DGE within cluster')
  return(list(counts=counts, conserved_markers=consmarkers, dge=dge))
}
```


## 0. NT/CM CD8+ T cells

```{r}
results <- per_cluster_DGE(0)
results$counts
results$conserved_markers
results$dge
```


## 4. EM CD8+ T cells

```{r}
results <- per_cluster_DGE(4)
results$counts
results$conserved_markers
results$dge
```

## 3. Neutrophils

```{r}
results <- per_cluster_DGE(3)
results$counts
results$conserved_markers
results$dge
```

## 7. Neutrophils

```{r}
results <- per_cluster_DGE(7)
results$counts
results$conserved_markers
results$dge
```

## 10. Neutrophils/eosinophils (?)

```{r}
results <- per_cluster_DGE(10)
results$counts
results$conserved_markers
results$dge
```

## 6. Plasmacytoid DC(?)

```{r}
results <- per_cluster_DGE(6)
results$counts
results$conserved_markers
results$dge
```